cdap-ui/app/cdap/components/AbstractWidget/SqlConditionsWidget/index.tsx

```diff
@@ -0,0 +1,291 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classnames from 'classnames';
+
+import withStyles, { WithStyles } from '@material-ui/core/styles/withStyles';
+
+import ThemeWrapper from 'components/ThemeWrapper';
+import If from 'components/If';
+import Rule from 'components/AbstractWidget/SqlConditionsWidget/Rule';
+
+export const styles = () => {
+  return {
+    root: {
+      width: '100%',
+    },
+    emptyMessage: {
+      marginBottom: 0,
+    },
+    textDanger: { 'white-space': 'pre-line' },
+    rulesContainer: {
+      display: 'flex',
+      width: '100%',
+      'flex-direction': 'column',
+    },
+  };
+};
+
+export interface IStage {
+  fieldName: string;
+  stageName: string;
+}
+
+export interface IRule extends Array<IStage> {}
+
+interface IRules extends Array<IRule> {}
+
+export interface IInputSchema {
+  [key: string]: string[];
+}
+
+interface ISqlConditionsWidgetProps extends WithStyles<typeof styles> {
+  value: string;
+  inputSchema: any[];
+  onChange: (arg0: any) => void;
+  disabled: boolean;
+}
+
+interface ISqlConditionsWidgetState {
+  warning: string;
+  error: string;
+  stageList: string[];
+  rules: IRules;
+  mapInputSchema: IInputSchema;
+}
+
+class SqlConditionsWidgetView extends React.Component<
+  ISqlConditionsWidgetProps,
+  ISqlConditionsWidgetState
+> {
+  public state = {
+    warning: null,
+    error: null,
+    stageList: [],
+    rules: [],
+    mapInputSchema: {},
+  };
+
+  public checkRulesForValidStageNames = () => {
+    const invalidRule = /[&\.=]/g;
+    let invalidStageNames = [];
+    this.state.rules.forEach((rule) => {
```

Comment:
Can this be simplified as,

```
this.state.rules.forEach((rule) => {
  rule.filter((field) => invalidRule.test(field.stageName))
  .map(stage => stage.stageName)
  .forEach(stage => invalidStageNames.push(stage));
});
```
This way for each stage, you iterate through the rules and for all those rules that has invalid stage names will get passed down to the map which maps to the actual stage name which is then pushed to `invalidStageNames` array. Does that look like a workable solution easier to read?

Reply:
Agreed, made changes.

Reply:
you don't need the forEach either. 
```
invalidStageNames = rule.filter().map()
```